[com_fields] Field type Repeatable. Editor and Textarea subfields remove HTML since J 3.9.7

ReLater opened the pull request:
Pull Request for Issue https://github.com/joomla/joomla-cms/issues/25187

### Summary of Changes
- Since Joomla 3.9.7 `subform` child fields are validated. So a minimum validation happens like known from "normal" form fields if no `filter` attribute is set for them.
- Because the com_fileds `repeatable` field uses a `subform` HTML tags are removed e.g. from editor or textarea contents during validation. This PR adds a filter field to any subfield row of `repeatable` fields.

### Testing Instructions
- Create a new com_fields field for Articles of type `repeatable` .
- Add a subfield of type Editor.
- Edit an article and enter some paragraphs in that editor.
- Save the article => paragraphs removed.

- Apply patch.
- Open above field.
- You'll see a new filter field for any row but for rows with type `Media` or `Number`:

![12-06-_2019_15-48-04](https://user-images.githubusercontent.com/20780646/59357130-8c177080-8d2a-11e9-9916-785b76da2ad3.jpg)

- Set Filter "Text" or "Safe Html."
- Save field.
- Open and edit article again or use a new one and enter some paragraphs => Save article  => paragraphs NOT removed.

- Test if filters are applied and work as expected for any field types.
- The filter options change depending on your selection in Type

ReLater committed:
Add field "fieldfilter" for subform fields filtering

ReLater committed:
Add filterfield

brianteeman commented:
Wasn't this as a result of JSST @snipersister @zero-24

ReLater commented:
> Wasn't this as a result of JSST @SniperSister @zero-24

Yes. See edited first comment.

alikon commented:
>How can I convert a draft pr to a regular one on GitHub????

just click on ready for review

![image](https://user-images.githubusercontent.com/181681/59359136-17463580-8d2e-11e9-8d9a-46eb1ecc578e.png)

ReLater commented:
> just click on ready for review

Thank you!

lunalars commented:
filter not making sense: maybe the filter should only be visible if editor or textarea is selected. 
added your code manually and also added showon="fieldtype:editor[OR]fieldtype:textarea to fieldfilter and did some quick tests. seems to work for all selected fieldtypes.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25189">issues.joomla.org/tracker/joomla-cms/25189</a>.</sub>

ReLater committed:
Showon to fieldfilter + options. Add more filters

ReLater commented:
New commit contains:
- Added more filter options. Taken over from single field types `text` and `textarea`.
- Added showon to `filterfield` (hide if `media` or `number`)
- Added showon to options.

Code style OK??? I hate too long lines ;-)

ReLater committed:
shorten showon attributes

ReLater commented:
@BPBlueprint @lunalars 
Could you please test this pr and mark your tests unsuccessful or successful on https://issues.joomla.org/tracker/joomla-cms/25189 . It's easy to apply and remove the patch with extension "Joomla! Patch Tester".

Normally I lose interest after 1 or 2 months in my prs and fixing branch conflicts ;-) if I don't use features myself.

lunalars commented:
If I set fieldtype to "Editor" and filter to "Use settings from Plugin" the HTML tags are still removed - should it be like this? 
other settings work just fine.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25189">issues.joomla.org/tracker/joomla-cms/25189</a>.</sub>

lunalars commented:
Sorry, forgot to mention: in the Editor Plugin (fields) Filter is set to "Text", so tags should not be removed.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25189">issues.joomla.org/tracker/joomla-cms/25189</a>.</sub>

ReLater committed:
Undefined property:stdClass:$fieldfilter for media

Quy commented on plugins/fields/repeatable/repeatable.php:
```diff
@@ -65,6 +65,11 @@ public function onCustomFieldsPrepareDom($field, DOMElement $parent, JForm $form
 			$child->addAttribute('name', $formField->fieldname);
 			$child->addAttribute('type', $formField->fieldtype);
 			$child->addAttribute('readonly', $readonly);
+			
```
Remove tabs.

ReLater commented:
Yes, you're right.
Set in plugin "Fields - Editor" filter to "Text".
In custom repeatable field select Editor and set filter to "Use settings from plugin".
Setting is not respected.

Sorry, then I have to give up here without assistance of others.

ReLater committed:
cs Remove tabs

lunalars commented:
Can "Use settings from Plugin" be removed here? 
I think it's not clear wich settings are meant: from editor or repeatable plugin.

ReLater committed:
Remove global option from fieldfilter

ReLater commented:
> Can "Use settings from Plugin" be removed here?

Give it a try! ;-) I removed global setting "Use settings from plugin".

lunalars commented:
very nice :-)
One more: I get `SimpleXMLElement::addAttribute(): Attribute already exists in plugins/fields/repeatable/repeatable.php on line 72` 
child attribute is already set on line 66

lunalars commented:
the mentioned warning appears on editing an article

lunalars commented:
I have tested this item :white_check_mark: successfully on 1824077a014dcbe8340c5485a472b3c42d46c1ec<br /><br />With the latest code selected filters work as expected. Tested all filters for text, textarea and editor.

The warning was gone after applying the latest patch correctly.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25189">issues.joomla.org/tracker/joomla-cms/25189</a>.</sub>

BPBlueprint commented:
I have tested this item ✅ successfully.

I have tested text, textarea and editor with and without filters.
Works as expected.

franz-wohlkoenig commented:
Status "Ready To Commit".

HLeithner commented:
thx

ReLater commented:
Thank you, guys!

HLeithner commented on plugins/fields/repeatable/repeatable.php:
```diff
@@ -65,6 +65,11 @@ public function onCustomFieldsPrepareDom($field, DOMElement $parent, JForm $form
 			$child->addAttribute('type', $formField->fieldtype);
 			$child->addAttribute('readonly', $readonly);
+
+			if (isset($formField->fieldfilter))
```
@ReLater 
As @bembelimen told me there seems to be to problems in this PR, first why do you set JNO to "0"?
And isset($formField->fieldfilter) will in this case always be true.

I would suggest to set JNO to value="" and change isset() to !empty() also I think you have to cast it to (string) and maybe better compare it to !== ''.

ReLater commented:
> And isset($formField->fieldfilter) will in this case always be true.

No. If you select type Media then fieldfilter is not set/is undefined and that was the reason for this line because you get a warning.

JNO=0 comes from the params that you can find in the params.xml of the plugins for single fields. E.g.
https://github.com/joomla/joomla-cms/blob/staging/plugins/fields/textarea/params/textarea.xml#L40

For me the question is: What effect should filter="0" have? Really no filtering (="unset"), fall back to "string" ... ? I don't know.

value="" was removed because it doesn't work. You can test that by setting a filter in the com_fields editor plugin. It's ignored in the repeatable field if you set "COM_FIELDS_FIELD_USE_GLOBAL".
https://github.com/joomla/joomla-cms/blob/staging/plugins/fields/textarea/params/textarea.xml#L39

TBH I never use com_fields and just provided this pr because we have submitted the security issue then and forgot to test com_fields repeatable and wanted to help concerned users quickly. @bembelimen can provide and describe a corrective pr if he thinks that something doesn't work here as expected.

HLeithner commented:
That's really strange but if it works it's ok for me. thx
